### PR TITLE
Wait until all indexing tasks are finished before switching to the new alias

### LINKED ISSUE #9811: Reindex is switching to the new alias too soon
STR:
- Trigger a reindex

Expected results:
- While the reindex is ongoing, the old index should still be present, and be used for search - the alias we use should not point to the new index yet.

Actual results:
- While the reindex is ongoing, we seem to be switching over to the new index immediately, removing the old one. The command returns fairly quickly showing the new alias configuration.

For users this means their search results will be incomplete for a while, possibly completely empty while we reindex, which is not cool.

### PR DESCRIPTION
**Pull request by diox**
This forces us to refactor the way we deal with reindexing to properly set up chains & groups together.

Fixes #9811

### CODE REVIEW

File: src/olympia/addons/indexers.py [hunk 1]
```diff
@@ -388,3 +393,120 @@ def extract_document(cls, obj):
                     cls.extract_field_raw_translations(preview, 'caption'))
 
         return data
+
```
Comment (diox) [inline src/olympia/addons/indexers.py]: Everything here is copied from `src/olympia/search/indexers.py` which I deleted. `reindex_tasks_group()` is the ne and improved `reindex()` which instead of firing reindex tasks directly, just returns a group of signatures via `create_chunked_tasks_signatures()`.

File: src/olympia/stats/management/commands/index_stats.py [hunk 2]
```diff
@@ -76,24 +81,20 @@ def index_stats(index, addons=None, dates=None):
                 continue
 
             num_days = (limits['max'] - limits['min']).days
-            today = date.today()
```
Comment (diox) [inline src/olympia/stats/management/commands/index_stats.py]: This command was "broken" in 3 different ways:
- Using `today` instead of the `max` date as the starting point
- Overlapping the `date_range`s in each iteration by one day
- Triggering tasks even if there was no data for that range 

None of these points terribly matter in production, where we're pretty much guaranteed to have data for every day until today, but in tests and locally it was spawning hundreds of useless empty tasks.

File: src/olympia/addons/indexers.py [hunk 3]
```diff
@@ -388,3 +393,120 @@ def extract_document(cls, obj):
...
+            },
+            'dict': {
+                # This filter is also useful for add-on names that have
+                # multiple words sticked together, but without a pattern that
+                # we can automatically recognize. To deal with those, we use
+                # a small dictionary of common words. It allows us to index
+                # "awesometabpassword"  as "awesome tab password", helping
+                # users looking for "tab password" find that add-on.
+                'type': 'dictionary_decompounder',
+                'word_list': [
```
Comment (eviljeff) [inline src/olympia/addons/indexers.py]: these are all English words, right?  How does it work with non-English?

File: src/olympia/addons/indexers.py [hunk 4]
```diff
@@ -388,3 +393,120 @@ def extract_document(cls, obj):
...
+
+    Intended to be used by reindexation (and tests), generally a bad idea to
+    call manually.
+    """
+    if index_name is None:
+        index_name = AddonIndexer.get_index_alias()
+
+    index_settings = copy.deepcopy(INDEX_SETTINGS)
+
+    if waffle.switch_is_active('es-use-classic-similarity'):
```
Comment (eviljeff) [inline src/olympia/addons/indexers.py]: Do we need this waffle flag?  classic is either better or not, I'd guess.

File: src/olympia/lib/es/management/commands/reindex.py [hunk 5]
```diff
@@ -95,6 +80,17 @@ def unflag_database():
     unflag_reindexing_amo()
 
 
+# FIXME: time limits, need to pass
```
Comment (eviljeff) [inline src/olympia/lib/es/management/commands/reindex.py]: follow-up issue or something you meant to fix in this issue?

File: src/olympia/addons/indexers.py (hunk 3 above)
Comment (diox) [reply src/olympia/addons/indexers.py]: It doesn't, this only really helps english - but we also decompose words automatically according to various criteria (CamelCase, any non-letter, etc)

File: src/olympia/lib/es/management/commands/reindex.py (hunk 5 above)
Comment (diox) [reply src/olympia/lib/es/management/commands/reindex.py]: I forgot to fix it here... but I'm not sure it matters - I don't think we need a specific limit. I'm inclined to remove it.

File: src/olympia/addons/indexers.py (hunk 4 above)
Comment (diox) [reply src/olympia/addons/indexers.py]: The waffle is currently on in prod. It has helped our relevancy problems a bit, but eventually it will need to be removed as it's deprecated in future versions of ES.

Evaluating whether or not we can remove it soon is on my TODO list - We'll see once I've done all the work I want to do for this quarter.

(In any case, I'm just moving the contents of another file here, it's not new code)

File: src/olympia/lib/es/management/commands/reindex.py (hunk 5 above)
Comment (diox) [reply src/olympia/lib/es/management/commands/reindex.py]: We currently have in our settings:
```python
CELERY_TIME_LIMITS = {
    # The reindex management command can take up to 3 hours to run.
    'olympia.lib.es.management.commands.reindex': {
        'soft': 10800, 'hard': 14400},
}
```

We know from last week it's not true - it may have been in the past, but now that the users, collections and compat data have been removed, it's way faster. 

More importantly, I don't think this can work now that the global indexing task has been replaced with a group - if we apply a limit to a group, it will be applied to each individual task, but not to the group itself (the doc say any options pass will be `applied to all tasks in the group`)

TL;DR: I'll remove this.

Comment (diox) [reply src/olympia/lib/es/management/commands/reindex.py]: 1a3a8eebe52b88cd445f2ae798390a861783b0c3
